golang/cosmos/x/vstorage/keeper/keeper.go

```diff
@@ -129,7 +137,28 @@ func (k Keeper) ImportStorage(ctx sdk.Context, entries []*types.DataEntry) {
 	for _, entry := range entries {
 		// This set does the bookkeeping for us in case the entries aren't a
 		// complete tree.
-		k.SetStorage(ctx, entry.Path, entry.Value)
+		k.SetStorage(ctx, types.StorageEntry{entry.Path, entry.Value})
+	}
+}
+
+func (k Keeper) MigrateNoDataPlaceholders(ctx sdk.Context) {
+	store := ctx.KVStore(k.storeKey)
+
+	iterator := sdk.KVStorePrefixIterator(store, nil)
+	emptyKeys := []*([]byte){}
+	{
+		defer iterator.Close()
+		for ; iterator.Valid(); iterator.Next() {
+			rawValue := iterator.Value()
+			if len(rawValue) == 1 && rawValue[0] == types.EncodedDataPrefix[0] {
+				key := iterator.Key()
+				emptyKeys = append(emptyKeys, &key)
+			}
+		}
+	}
+
+	for _, key := range emptyKeys {
+		store.Set(*key, types.EncodedNoDataValue)
 	}
```

Comment:
Did you have a reason not to avoid allocating emptyKeys and do this within the iteration loop?

Reply:
From what I gather you cannot mutate the store while iterating it.

Reply:
https://github.com/cosmos/cosmos-sdk/blob/47f46643affd7ec7978329c42bac47275ac7e1cc/store/types/store.go#L224-L226

```
	// CONTRACT: No writes may happen within a domain while an iterator exists over it.
	// Exceptionally allowed for cachekv.Store, safe to write in the modules.
	Iterator(start, end []byte) Iterator
```

Reply:
Added a comment explaining why the 2 stage process.